src/main/java/reposense/authorship/AuthorshipReporter.java

```diff
@@ -37,4 +60,28 @@ public static AuthorshipSummary generateAuthorshipSummary(RepoConfiguration conf
         return FileResultAggregator.aggregateFileResult(fileResults, config.getAuthorList(),
                 config.getAllFileTypes());
     }
+
+    /**
+     * Returns the number of lines in the file represented by {@code textFileInfo}.
+     */
+    private static int getNumLinesInFile(RepoConfiguration config, FileInfo textFileInfo) {
+        BufferedReader reader;
+        String path = Paths.get(FileUtil.getRepoParentFolder(config).toString(), config.getRepoName(),
+                textFileInfo.getPath().replace("/", "\\")).toString();
+        try {
+            reader = new BufferedReader(new FileReader(path));
+        } catch (FileNotFoundException fnfe) {
+            return 0;
+        }
+        int lines = 0;
+        try {
+            while (reader.readLine() != null) {
+                lines++;
+            }
```

Comment:
This may be quite time consuming as it loads every file once to count the number of lines. Instead, I found [a way to count total number of lines in Git](https://stackoverflow.com/questions/4822471/count-number-of-lines-in-a-git-repository/4822890#4822890). It uses the following command:

```shell
git diff --shortstat `git hash-object -t tree /dev/null`
```

It's a Unix command but I think it should be possible to generalize to Windows as well. Using the [servo](https://github.com/servo/servo) repository, it seems to run reasonably fast:

```shell
$ time git diff --shortstat `git hash-object -t tree /dev/null`
 101260 files changed, 6570287 insertions(+)

real    0m2.601s
user    0m1.887s
sys     0m0.715s
```

Perhaps we can consider using this method instead?

Reply:
I tried implementing this `git diff` based method in RepoSense. I was able to get a Windows-compatible version by splitting this into 2 commands, first using a `git hash-object` command to get the hash and then running `git diff`.

I timed both methods within Java using `System.nanotime()`. Based on this, it seems like the `git diff` method is much slower - on my Windows 10 machine, on the servo repository, `git diff` took around 20 seconds on average, compared to less than 1 sec using the current method of reading every file.

Additionally, the `git diff` method counts the lines in every single file in the repository. However, the current method only looks at the files that are going to be analyzed for the authorship report. Since users can filter the files that should be included in the authorship report (e.g. based on the file format) I feel that the current method would provide a more accurate count.

Reply:
I think it is a good point that there may be some files that are filtered, but it is provided that users specify the file globs to ignore. 

> I timed both methods within Java using `System.nanotime()`. Based on this, it seems like the `git diff` method is much slower - on my Windows 10 machine, on the servo repository, `git diff` took around 20 seconds on average, compared to less than 1 sec using the current method of reading every file.

On this point, I tried to reproduce this on my Windows machine using PowerShell and on Intellij (with SSD storage). I got the following output for PowerShell:

```powershell
PS C:\Users\user\Documents\servo> Measure-Command {git diff --shortstat 4b825dc642cb6eb9a060e54bf8d69288fbee4904}


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 10
Milliseconds      : 965
Ticks             : 109652154
TotalDays         : 0.000126912215277778
TotalHours        : 0.00304589316666667
TotalMinutes      : 0.18275359
TotalSeconds      : 10.9652154
TotalMilliseconds : 10965.2154
```

I took a snippet of the output log from Intellij:
```
03:31:27 - [7/8] Analyzing https://github.com/servo/servo.git (master)...
03:32:39 - https://github.com/servo/servo.git (master) does not contain a standalone config file.
03:32:39 - https://github.com/servo/servo.git (master) has no authors specified, using all authors by default.
03:32:41 - [WARNING] Duplicate alias camelid found. The alias will belong to the last author - camelid
03:32:41 - [WARNING] Duplicate alias camelid@users.noreply.github.com found. The alias will belong to the last author - camelid
03:32:41 - [WARNING] Duplicate alias cybai found. The alias will belong to the last author - cybai
03:32:41 - [WARNING] Duplicate alias cybai@users.noreply.github.com found. The alias will belong to the last author - cybai
03:32:41 - Extracting commits info for https://github.com/servo/servo.git (master)...
03:50:00 - Analyzing commits info for https://github.com/servo/servo.git (master)...
03:50:04 - Extracting relevant file info from https://github.com/servo/servo.git (master)...
03:50:52 - [WARNING] There are a large number (97401) of text files to be analyzed, comprising a total of 6571182 lines. As such, RepoSense may take a long time to generate the report. Performance may be improved by setting values for the "File formats" and  "Ignore Glob List" columns in the repo-config.csv so as to reduce the number of files to be  analyzed.
```

The code on Intellij is running the branch for this PR, and it takes about 48 seconds compared to the `git diff` method of about 10 seconds. Can I check how you are able to get less than 1 second using this method that you are proposing?

Reply:
I think the 48 seconds figure is not accurate. The "Extracting relevant file info..." log appears at the beginning of the `extractTextFileInfos` method in `FileInfoExtractor`, which means that the 48 seconds includes all the time taken to generate the whole list of `FileInfo` objects.

To get a more accurate measurement, I put a couple of logs at the start & end of both the `git diff` and the function that reads the files.

```
05:22:14 - [1/1] Analyzing https://github.com/servo/servo.git (master)...
05:24:00 - https://github.com/servo/servo.git (master) does not contain a standalone config file.
05:24:00 - https://github.com/servo/servo.git (master) has no authors specified, using all authors by default.
05:24:00 - Extracting commits info for https://github.com/servo/servo.git (master)...
05:24:36 - Analyzing commits info for https://github.com/servo/servo.git (master)...
05:24:37 - Extracting relevant file info from https://github.com/servo/servo.git (master)...
05:24:48 - [WARNING] "tests/wpt/web-platform-tests/origin-policy/policies/op10 cspfp-valid.json      " is an invalid file path. Skipping this directory.
05:24:48 - [WARNING] "tests/wpt/web-platform-tests/origin-policy/policies/op2 cspfp-double-top-level.json    " is an invalid file path. Skipping this directory.
05:24:48 - [WARNING] "tests/wpt/web-platform-tests/origin-policy/policies/op5 cspfp-non-object.json  " is an invalid file path. Skipping this directory.
05:24:48 - [WARNING] "tests/wpt/web-platform-tests/origin-policy/policies/op6 cspfp-non-string.json  " is an invalid file path. Skipping this directory.
05:24:49 - Starting GitDiff
05:25:06 - GitDiff complete, 6570263 lines found
05:25:06 - Starting to read files
05:25:07 - Reading files complete, 1701913 lines found
05:25:07 - [WARNING] There are a large number (11584) of text files to be analyzed, comprising a total of 1701913 lines. As such, RepoSense may take a long time to generate the report. Performance may be improved by setting values for the "File formats" and  "Ignore Glob List" columns in the repo-config.csv so as to reduce the number of files to be  analyzed.
```

Based on this, the `git diff` method took about 17 sec, and the other method took about 1 sec.